instrumentation/kafka-clients/kafka-clients-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTracing.java

```diff
@@ -0,0 +1,118 @@
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersGetter;
+import io.opentelemetry.instrumentation.kafka.KafkaHeadersSetter;
+import io.opentelemetry.instrumentation.kafka.KafkaSingletons;
+import java.util.Objects;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.header.Headers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KafkaTracing {
+
+  private static final Logger logger = LoggerFactory.getLogger(KafkaTracing.class);
+
+  private static final TextMapGetter<Headers> GETTER = new KafkaHeadersGetter();
+
+  private static final TextMapSetter<Headers> SETTER = new KafkaHeadersSetter();
+
+  private Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter =
+      KafkaSingletons.producerInstrumenter();
+  private Instrumenter<ConsumerRecord<?, ?>, Void> consumerInstrumenter =
+      KafkaSingletons.consumerProcessInstrumenter();
+
+  KafkaTracing() {}
+
+  public static KafkaTracingBuilder newBuilder(OpenTelemetry openTelemetry) {
+    return new KafkaTracingBuilder(openTelemetry);
+  }
+
+  private static TextMapPropagator propagator() {
+    return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+  }
+
+  void setProducerInstrumenter(Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter) {
+    this.producerInstrumenter = Objects.requireNonNull(producerInstrumenter);
+  }
+
+  void setConsumerInstrumenter(Instrumenter<ConsumerRecord<?, ?>, Void> consumerInstrumenter) {
+    this.consumerInstrumenter = Objects.requireNonNull(consumerInstrumenter);
+  }
+
+  /**
+   * Extract Context from record headers.
+   *
+   * @param headers record headers
+   * @return context
+   */
+  private static Context extractContext(Headers headers) {
+    return propagator().extract(Context.current(), headers, GETTER);
+  }
+
+  /**
+   * Inject current Context to record headers.
+   *
+   * @param headers record headers
+   */
+  private static void inject(Headers headers) {
+    inject(Context.current(), headers);
+  }
+
+  /**
+   * Inject Context to record headers.
+   *
+   * @param context the context
+   * @param headers record headers
+   */
+  private static void inject(Context context, Headers headers) {
+    propagator().inject(context, headers, SETTER);
+  }
+
+  /**
+   * Build and inject span into record. Return Runnable handle to end the current span.
+   *
+   * @param record the producer record to inject span info.
+   * @return runnable to close the current span
+   */
+  public <K, V> Runnable buildAndInjectSpan(ProducerRecord<K, V> record) {
+    Context spanContext = extractContext(record.headers());
```

Comment:
You should use the correct context when sending the message - the context is supposed to be injected into the producer records headers so that it's transmitted across the "Kafka barrier"; but you should start the span from the current span so that this instrumentation correctly starts the span within the current trace.

Reply:
So I should just use Context::current() ?

Reply:
Yep!